File: api/src/test/java/net/blueshell/api/factory/model/SurveyFactory.java
package net.blueshell.api.factory.model;

import net.blueshell.api.testutil.ModelTestUtils;
import com.github.javafaker.Faker;
import lombok.RequiredArgsConstructor;
import net.blueshell.api.model.survey.Survey;
import org.springframework.stereotype.Component;

import java.util.concurrent.atomic.AtomicLong;

/**
 * Factory for Survey model test instances.
 */
@Component
@RequiredArgsConstructor
public class SurveyFactory {

    private static final AtomicLong counter = new AtomicLong(1000);
    private final Faker faker;
    private final QuestionFactory questionFactory;

    public Survey createBasic() {
        Survey survey = new Survey();
        survey.setResponseCount(0L);
        return survey;
    }

    public Survey createFull() {
        Survey survey = createBasic();
        for (int i = 0; i < faker.number().numberBetween(3, 8); i++) {
            survey.getQuestions().add(questionFactory.createForSurvey(survey));
        }
        return survey;
    }

    public Survey createWithCustomizations(java.util.function.Consumer<Survey> customizer) {
        Survey survey = createFull();
        customizer.accept(survey);
        return survey;
    }

    public Survey createWithQuestions(int count) {
        return createWithCustomizations(survey -> {
            survey.getQuestions().clear();
            for (int i = 0; i < count; i++) {
                survey.getQuestions().add(questionFactory.createForSurvey(survey));
            }
        });
    }

    private Long generateId() {
        return counter.incrementAndGet();
    }
}

File: api/src/test/kotlin/net/blueshell/api/factory/model/survey/SurveyFactory.kt
package net.blueshell.api.factory.model.survey

import com.github.javafaker.Faker
import net.blueshell.api.survey.persistence.Question
import net.blueshell.api.survey.persistence.Survey
import org.springframework.stereotype.Component
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Consumer

/**
 * Factory for Survey model test instances.
 */
@Component
class SurveyFactory(
    private val faker: Faker,
    private val questionFactory: QuestionFactory
) {

    fun createBasic(): Survey {
        val survey = Survey()
        survey.responseCount = 0L
        return survey
    }

    fun createFull(): Survey {
        val survey = createBasic()
        repeat(faker.number().numberBetween(3, 8)) {
            survey.addQuestion(questionFactory.createForSurvey(survey))
        }
        return survey
    }

    fun createWithCustomizations(customizer: Consumer<Survey>): Survey {
        val survey = createFull()
        customizer.accept(survey)
        return survey
    }

    fun createWithQuestions(count: Int): Survey {
        return createWithCustomizations { survey ->
            val questions = mutableListOf<Question>()
            repeat(count) {
                questions.add(questionFactory.createFull())
            }
            survey.replaceQuestions(questions)
        }
    }

    private fun generateId(): Long = COUNTER.incrementAndGet()

    private companion object {
        val COUNTER = AtomicLong(1000)
    }
}
